add keepalived resources, specs, tests, and docs, and modularize recipes

nathwill opened the pull request:
- [x] remove attribute-driven configuration
- [x] modularize recipes (helps wrapper-cookbooks pick and choose)
- [x] add resources
- [x] add provider
- [x] changelog update
- [x] version bump
- [x] specs for new recipes
- [x] specs for new resources
- [x] specs for new provider
- [x] test suite for new recipes/resources
- [x] documentation updates

nathwill commented:
Alright, this is ready for review. Apologies for the monstrous PR, it all just kinda hung together once the groundwork was in place...

This should be 100% backwards-compatible, though :cake:. It's set up to let users opt into the new configuration via the `node['keepalived']['use_attributes']` attribute, which defaults to current functionality, but, if disabled, the configure recipe will set up an include in `keepalive.conf` to load resource-generated configs.

tas50 commented:
@nathwill Thanks for taking this one on. I'm looking forward to the review and seeing this move to a nice composable resource driven design.

nathwill commented:
@tas50 :+1: me too! i just updated the docs to reference the upstream source doc directory, since it seems to be a good deal more complete than the man pages or docs on the official site, and added in the vrrp_script resource (super important!), which for some reason is missing from all of the man pages i read, but was included in the upstream synopsis doc. i also backfilled the specs and property documentation for the directives that were in the upstream docs, so i'm like 95% sure at this point that we've got good, complete coverage.

tas50 commented on .rubocop.yml:
```diff
@@ -1,31 +1,12 @@
-  Enabled: false
-CyclomaticComplexity:
-  Enabled: false
+    - spec/**/*
```
If possible we'd like to keep the rubocop configs on the test/spec dirs.  It's nice that they're consistent and readable as well.

tas50 commented on .kitchen.yml:
```diff
@@ -6,13 +6,15 @@ provisioner:
     run_list: apt::default
-  - name: fedora-21
   - name: fedora-22
+    run_list: test::dnf
```
We added a DNF backwards compat recipe to the last version of the yum cookbook.  It might make more sense to use that here similar to how we use apt::default

nathwill replied:
oh rad, i hadn't seen that; updated accordingly.

tas50 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@ keepalived Cookbook CHANGELOG
 =============================
 This file is used to list changes made in each version of the keepalived cookbook.
 
+v1.4.0 (2016-03-??)
```
This should be versioned UNRELEASED for now

tas50 commented on metadata.rb:
```diff
@@ -4,9 +4,12 @@
 description       'Installs and configures keepalived'
 long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
-version           '1.3.0'
+version           '1.4.0'
```
We bump these at the time of release so this will need to stay at 1.3.0

tas50 commented on recipes/configure.rb:
```diff
@@ -0,0 +1,64 @@
+  directory include_path do
+    owner 'root'
+    group 'root'
+    mode 0755
```
This should be either 00755 or '0755' to preserve the leading 0

tas50 replied:
I just realized it was wrong in our old code.  NM that

nathwill commented:
test failure looks like a stalled build on travis, this passed locally

nathwill commented:
@tas50 anything else you need here before it's ready for review?

tas50 commented:
Property is the new hotness, but it might give you some namespacing issues so be careful.

nathwill commented:
> it might give you some namespacing issues so be careful.

i'm dying laughing over here... i was told to avoid LWRPBase due to namespacing issues at the same time i was told property's the new preference for this stuff... apparently there is no solution that isn't fraught with peril? >__<. Chef, what hast thou wrought!?!?

tas50 commented:
I moved over to the sweet new 12.5 resources and everything was great until the namespace stuff caused me grief.  Namespacing is hard apparently.  BTW: I will have a real review for you soon.  I've been buried in AWS land for a while.

nathwill commented:
okey doke, sounds good; this is obviously a fairly significant change, so i wasn't expecting overnight sign-off,and the pause has given me enough time to look back over it and make a few minor improvements :smile:.

btw, i didn't know what to do about the keepalived_test cookbook. it looks like at some point in the past it was intended to exercise the attribute-driven config, but was unused on master when i started on the resource stuff. i added some integration tests for the keepalived cookbook's default behavior, since there weren't any, but if we want to exercise more of the attr-driven behavior via that cookbook, i'm happy to help tackle that (maybe via a second PR once this is merged).

nathwill commented:
any update @tas50 ?

![waiting](http://www.reactiongifs.us/wp-content/uploads/2013/10/waiting_game_of_thrones.gif)

tas50 commented:
So in general I'm trusting your knowledge of keepalived since I've never used it.  I'm sure that part is all fantastic.  The only real feedback is I have is more of a question.  Is there a reason to use the use_attributes attribute?  It seems like the previous attribute only implementation was limiting to the point that most people would probably fork this cookbook. We should just break compatibility with a major version number and make it a resource only cookbook.  It gives a consistent interface that is inline with our other resource only cookbooks. Setting an attribute to use resources is a bit odd.

nathwill commented:
@tas50 sure, i'm 100% OK with that (and agree attributes are extremely limiting), I just wanted to provide backwards compatibility by default since this isn't my cookbook and it was pretty easy to do. If you prefer to flip 100% to resource-driven, I can update this accordingly.

tas50 commented:
Break the world.  It's worth it for the proper design

nathwill commented:
:+1: sounds great, i'll tear out the old code and make sure this is still in shape then.

> I'm trusting your knowledge of keepalived since I've never used it

i've no idea what the policies are around this kind of thing, but fwiw, i'd be happy to help maintain this cookbook going forward; we use keepalived extensively, so i've no intention of this being a drive-by overhaul and then forgetting about it :D

tas50 commented:
@nathwill I was going to ask you if you wanted to be a maintainer on this. I'll get that setup.

nathwill commented:
@tas50 great, thanks! i did a bunch of cleanup on the code to pull out the old attribute-driven stuff; i want to make another sweep through the readme and add some more verbose documentation, but after that i think this is ready to go!
